postgres-protocol/src/escape/mod.rs

```diff
@@ -0,0 +1,93 @@
+//! Provides functions for escaping literals and identifiers for use
+//! in SQL queries.
+//!
+//! Prefer parameterized queries where possible; see
+//! [`Client::query`](crate::Client::query). Do not escape parameters.
+
+#[cfg(test)]
+mod test;
+
+/// Escape a literal and surround result with single quotes. Not
+/// recommended in most cases.
+///
+/// If input contains backslashes, result will be of the form `
+/// E'...'` so it is safe to use regardless of the setting of
+/// standard_conforming_strings.
+pub fn escape_literal(input: &str) -> String {
```

Comment:
Can we have these return a `Cow<str>` so the no escaping needed case doesn't need to reallocate?

Reply:
`escape_literal()` follows the semantics of `PQescapeLiteral()`, which always allocates because the result must be surrounded by quote marks. This is different from `PQescapeStringConn()` which has some messy history that I don't think we need to follow because we should be using parameters most of the time anyway. In those cases where we still need it (like utility commands or replication commands), either form will suffice.

We could potentially optimize `escape_identifier()`, but it would deviate from libpq. I don't see this as likely to be a hot path, so the trade-off doesn't seem great to me.

Reply:
Ah yeah I missed that quotes were always added. The behavior here makes sense in that case.